### PR TITLE
324 Reset State on Logout

### LINKED ISSUE #324: Reset state on logout
State should be reset when logging out, currently it's inconsistent. Routing state, apollo data, etc

### PR DESCRIPTION
**Pull request by richarddubay**
Fixes #324 

## Creator

- [ ] Build relevant tests if any apply
- [ ] Ensure there are no new warnings
- [ ] Upload an animated GIF showcasing the solution (if it applies)
- [ ] Set a relevant reviewer

## Reviewer

- [ ] Run `test` and make sure all tests pass
- [ ] Review function and form on iOS
- [ ] Review function and form on Android
- [ ] Review function and form on Web
- [ ] Review new test logic

### CODE REVIEW

**Comment by richarddubay** [comment]
@burkeshartsis @IsaacHardy This looks like it's working, but I'm pretty concerned about the flashing that occurs when the app is reset. It happens both on native and web. Is there anything I can do to make that _not_ happen?

**Comment by richarddubay** [comment]
I could probably just use some direction on what I'm looking for to prove that this is even working. Looking at the `cache` tab in the Apollo Dev Tools, when I start up the web, there is nothing in the cache for `currentPerson`. When I sign in the `currentPerson` gets set with data. When I then log out, the `currentPerson` still shows data in the cache.

But conversely, before signing in, if I run the `currentPerson` query, it returns null. Then after I sign in the query returns my person data. After I log out, the query returns null again. 

So if the Apollo Dev Tools can be believed, the `window.location.reload('true');` function doesn't clear out the cache when the page is reloaded. FWIW, if I just refresh the page manually there is no change in the cache either. I have to physically close the tab for the cache to go away.

I'm not sure what I should be looking at as the source of truth that what I'm seeing is either right or wrong. 

@IsaacHardy or @burkeshartsis ... Thoughts?

**Comment by Lepozepo** [comment]
We can use `Client.queryManager.dataStore.cache.data` to inspect state easier. I'm writing up a function that takes from [this](https://github.com/apollographql/apollo-client/blob/6eec674f206cda11cf667a89022354b1476f5503/packages/apollo-client/src/core/QueryManager.ts#L801) to clear the cache instead so we don't need to do any funky resets

**Comment by Lepozepo** [comment]
`Client.queryManager.dataStore.reset` <- this seems to be the magic sauce

**Comment by Lepozepo** [comment]
Here it is working on web, double checking native
newspring

**Comment by Lepozepo** [comment]
Seems to work on native but we're getting some noise from `fetchQueryPromises`, turning the whole thing into a promise should fix this.

**Comment by IsaacHardy** [comment]
@Lepozepo Is this ready for review?

**Comment by Lepozepo** [comment]
Ok guys, this should be ok. Locally everything seems relatively stable. I'm concerned about the router though. `ProtectedRoute` may be doing damage to the log out flow but I haven't been able to pin point exactly what is causing it.

**Comment by IsaacHardy** [comment]
@Lepozepo Is there a way to see how the `ProtectedRoute` is doing damage to the log out flow? I'd like to understand this a bit better.

**Comment by Lepozepo** [comment]
@IsaacHardy I'm not 100% certain yet (will look into it further between today and tomorrow) but it seems that after the logout action is triggered the current route reruns the logic in `ProtectedRoute` causing a redirect but the current function is also redirecting. Actually, now that I'm typing it out, maybe the solution is to remove the redirect from the `onPress` handler on `LogoutTouchable` as the logic for the redirect already exists.

**Comment by Lepozepo** [comment]
Or we would need to route to a placeholder route while logout is occurring and then redirect

**Comment by IsaacHardy** [comment]
@burkeshartsis Do you know what is still needed for this PR? The above comment from @Lepozepo leads me to believe more work is needed.

**Comment by burkeshartsis** [comment]
Definitely seems like there is more work here. @richarddubay or @IsaacHardy do you feel like you can pick up where @Lepozepo left off?

**Comment by richarddubay** [comment]
@Lepozepo I commented out that `goBackTo()` function call inside the `onPress` handler. Can you see if that takes care of what you were seeing was the problem? Interesting that the `goBackTo()` was trying to go to `/profile` but then was redirecting to `/login`. Any reason why that wouldn't just go back to the home feed?

**Comment by burkeshartsis** [comment]
The redirect was by design. In talking about it with @conrad-vanl, he and I decided the best UX was to send the user to the login screen (marginally better) when they logged out as opposed to the home screen.

**Comment by richarddubay** [comment]
@burkeshartsis So when this gets merged with the other modal styling changes maybe this will be better. Right now if you close the login modal you just end up on an empty profile page. 👍

**Comment by IsaacHardy** [comment]
@richarddubay This is labeled `ready for review`, am I good to go in reviewing? I felt that we were waiting on this for some reason.

**Comment by richarddubay** [comment]
@IsaacHardy I made the change that @Lepozepo mentioned above (to remove the goBackTo() route) so I figured that was enough to get this thing reviewed and moving again.

**Comment by burkeshartsis** [comment]
@IsaacHardy can you take over reviewing this. I'm not completely sure where to begin/pickup testing this.

File: src/settings/index.js [hunk 1]
```diff
@@ -18,10 +18,14 @@ export { ProfileDetails, ProfileAddress, ChangePassword } from './forms';
...
-      logout();
-      goBackTo({ to: '/', history, replace: true });
+  withProps(({ logout }) => ({
+    async onPress() {
+      await logout();
+      // TODO: There is something wrong with routing
+      // at this point. The user is properly logged out
+      // and cache is cleared before goBackTo runs
+      // RD: Commenting out the goBackTo per Marcelo comment on Issue #324.
+      // goBackTo({ to: '/profile', history, replace: true });
```
[the remark this answers is not in the document]
Comment (IsaacHardy) [reply src/settings/index.js]: @Lepozepo I'm not sure the correct solution, but when you test this change you will notice the behavior i mentioned above and why I don't believe it is something we can move forward with.

**Comment by Lepozepo** [comment]
Mmm, I think the nature of `branch` is causing this. When props change, `renderComponent` reruns so redirects are called multiple times. Need to dig a little bit more though, there's also a logical error where, if you're in settings and you log out, settings attempts a redirect (so we don't need to have a redirect function because the component knows to redirect the user) but it redirects to /login and sets the referrer to /settings which is blank because it is "protected". We just need to split the flows there :D

**Comment by Lepozepo** [comment]
All right guys, I feel pretty safe with this now that `ProtectedRoute` works as expected on web, just need to test it out on native

**Comment by Lepozepo** [comment]
Ok ^_^ the behavior of ProtectedRoute is consistent across platforms now. There seems to be noise in native though when you hit the close button (as now it's being redirected properly) and you've been redirected to login where you need to click it twice to close. This issue should be treated separately though. @burkeshartsis @IsaacHardy let me know what you guys think

**Pull request merged**